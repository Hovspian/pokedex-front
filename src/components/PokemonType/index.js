import React from 'react';
import '../../styles/components/PokemonType/PokemonType.css';
import {Badge} from 'reactstrap';

class PokemonType extends React.Component {
    render () {
        const self = this;
        return (
            <div>
                {this.props.types.map(function(type) {
                    return <Badge className={'type ' + type} key={self.props.id + '-' + type} color="primary" pill>{type}</Badge>
                })}
            </div>
        );
    }
}

export default PokemonType
