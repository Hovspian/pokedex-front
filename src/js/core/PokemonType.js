import React from 'react';
import PropTypes from 'prop-types';
import { Badge, Popover, PopoverBody } from 'reactstrap';

import { TYPE_LOOKUP } from '../core/Constants';
import '../../styles/core/PokemonType.css';

const propTypes = {
  types: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.arrayOf(
      PropTypes.shape({
        type: PropTypes.string,
        multiplier: PropTypes.string
      }))
  ]).isRequired,
  isCheckbox: PropTypes.bool,
  toggleType: PropTypes.func,
  isWeakness: PropTypes.bool,
  isLarge: PropTypes.bool,
  checkedTypes: PropTypes.arrayOf(PropTypes.bool)
};

class PokemonType extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: false,
    };
    this.toggleHover = this.toggleHover.bind(this);
  }

  toggleHover() {
    this.setState({ isOpen: !this.state.isOpen });
  }

  /**
   * Returns an element for the weakness with a popover if the multiplier is extra strong.
   * A multiplier is normally 2x
   *
   * @param {Object} weakness - the weakness being rendered
   * @param {String} weakness.type - the weakness's name
   * @param {String} weakness.multiplier - the weakness's multiplier
   */
  renderWithHover(weakness) {
    const multiplier = parseFloat(weakness.multiplier);
    if (multiplier > 2) {
      return (
        <span>
          {weakness.type}
          <span className="extra-damage"
                id={`extra-damage-${weakness.type}`}
                onMouseOver={this.toggleHover}
                onMouseOut={this.toggleHover}
          >
            *
          </span>
          <Popover placement="top"
                   isOpen={this.state.isOpen}
                   target={`extra-damage-${weakness.type}`}
                   hideArrow={true}
          >
            <PopoverBody>Deals {multiplier}x damage</PopoverBody>
          </Popover>
        </span>
      );
    } else {
      return <span>{weakness.type}</span>;
    }
  }

  shouldRenderChecked(checked, typeName) {
    if (!checked || !checked[TYPE_LOOKUP[typeName] - 1]) {
      return '';
    } else {
      return ' checked';
    }
  }

  shouldToggle(typeName) {
    if (this.props.isCheckbox) {
      this.props.toggleType(TYPE_LOOKUP[typeName] - 1, this.props.isWeakness);
    }
  }

  renderType(isLarge, isWeakness, type) {
    if (isLarge && isWeakness) {
      return (
        <h4>{this.renderWithHover(type)}</h4>
      );
    } else if (isLarge) {
      return (
        <h4>{type}</h4>
      );
    } else {
      return (
        <span>{type}</span>
      );
    }
  }

  render() {
    return (
      <div>
        {this.props.types.map((type, index) => {
          const typeName = this.props.isWeakness ? type.type : type;
          const typeClass = this.props.isLarge ? ' large-type' : '';
          const typeCheckbox = this.props.isCheckbox ? ' checkbox' : '';
          const typeChecked = this.shouldRenderChecked(this.props.checkedTypes, typeName);
          return (
            <Badge className={`type ${typeName}${typeClass}${typeCheckbox}${typeChecked}`}
                   key={typeName + index}
                   onClick={() => {this.shouldToggle(typeName)}}
                   pill
            >
              {this.renderType(this.props.isLarge, this.props.isWeakness, type)}
            </Badge>
          );
        })}
      </div>
    );
  }
}

PokemonType.propTypes = propTypes;

export default PokemonType;
