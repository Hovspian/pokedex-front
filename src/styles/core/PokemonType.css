.type {
  margin: 3px;
}

/* Colors taken from https://bulbapedia.bulbagarden.net/wiki/Type */
.bug {
  background-color: #A8B820;
}

.electric {
  background-color: #F8D030;
}

.dark {
  background-color: #705848;
}

.dragon {
  background-color: #7038F8;
}

.fairy {
  background-color: #EE99AC;
}

.fighting {
  background-color: #C03028;
}

.fire {
  background-color: #f08030;
}

.flying {
  background-color: #A890F0;
}

.ghost {
  background-color: #705898;
}

.grass {
  background-color: #78C850;
}

.ground {
  background-color: #E0C068;
}

.ice {
  background-color: #98D8D8;
}

.normal {
  background-color: #a8a878;
}

.poison {
  background-color: #A040A0;
}

.psychic {
  background-color: #F85888;
}

.rock {
  background-color: #B8A038;
}

.steel {
  background-color: #B8B8D0;
}

.water {
  background-color: #6890F0;
}

.bug-border {
  color: #A8B820;
}

.electric-border {
  color: #F8D030;
}

.dark-border {
  color: #705848;
}

.dragon-border {
  color: #7038F8;
}

.fairy-border {
  color: #EE99AC;
}

.fighting-border {
  color: #C03028;
}

.fire-border {
  color: #f08030;
}

.flying-border {
  color: #A890F0;
}

.ghost-border {
  color: #705898;
}

.grass-border {
  color: #78C850;
}

.ground-border {
  color: #E0C068;
}

.ice-border {
  color: #98D8D8;
}

.normal-border {
  color: #a8a878;
}

.poison-border {
  color: #A040A0;
}

.psychic-border {
  color: #F85888;
}

.rock-border {
  color: #B8A038;
}

.steel-border {
  color: #B8B8D0;
}

.water-border {
  color: #6890F0;
}

.large-type {
  width: 130px;
}

.checked {
  box-shadow: 0px 0px 10px 10px #ddd;
}

.checkbox {
  margin: 10px 0;
  cursor: pointer;
  user-select: none;
}

.extra-damage {
  margin-left: 5px;
  padding: 17px 6px 5px 6px;
  display: inline-block;
  border-radius: 50%;
  background-color: grey;
  vertical-align: sub;
  line-height: 0;
  position: relative;
  cursor: pointer;
}
