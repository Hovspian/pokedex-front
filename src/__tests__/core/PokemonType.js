import React from 'react';
import { shallow } from 'enzyme';

import PokemonType from '../../js/core/PokemonType';
import ExampleJSON from '../../../local_data/example_single.json';
import ExampleDetailJSON from '../../../local_data/example_detail.json';


describe('PokemonType render', () => {
  it('should render Type component without crashing', () => {
    shallow(<PokemonType isLarge isWeakness types={ExampleJSON.types} />);
  });

  it('should match Type component snapshot', () => {
    const wrapper = shallow(<PokemonType isLarge isWeakness types={ExampleJSON.types} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Tests render output for different types and sizes', () => {
  it('should render 4 types', () => {
    const wrapper = shallow(<PokemonType isLarge isWeakness types={ExampleDetailJSON.forms[0].weaknesses} />);
    expect(wrapper.children().length).toEqual(ExampleDetailJSON.forms[0].weaknesses.length);
  });

  it('should render a hover over', () => {
    const wrapper = shallow(<PokemonType isLarge isWeakness types={ExampleDetailJSON.forms[0].weaknesses} />);

    expect(wrapper.exists('.extra-damage')).toEqual(true);
  });

  it('should not render a hover over', () => {
    const example = Object.assign({}, ExampleDetailJSON);
    // remove x4 damage multiplier
    example.forms[0].weaknesses.splice(-1, 1);
    const wrapper = shallow(<PokemonType isLarge isWeakness types={example.forms[0].weaknesses} />);
    expect(wrapper.exists('.extra-damage')).toEqual(false);
  });

  it('should render a small type', () => {
    const wrapper = shallow(<PokemonType types={ExampleDetailJSON.forms[0].types} />);
    expect(wrapper.exists('h4')).toEqual(false);
  });

  it('should render a large type', () => {
    const wrapper = shallow(<PokemonType isLarge types={ExampleDetailJSON.forms[0].types} />);
    expect(wrapper.exists('h4')).toEqual(true);
  });

});
