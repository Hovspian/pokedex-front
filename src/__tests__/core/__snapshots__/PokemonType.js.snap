// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonType render should match Type component snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PokemonType
    isLarge={true}
    isWeakness={true}
    types={
        Array [
            "water",
            "flying",
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Badge
          className="type undefined large-type"
          color="secondary"
          onClick={[Function]}
          pill={true}
          tag="span"
>
          <h4>
                    <span />
          </h4>
</Badge>,
        <Badge
          className="type undefined large-type"
          color="secondary"
          onClick={[Function]}
          pill={true}
          tag="span"
>
          <h4>
                    <span />
          </h4>
</Badge>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "NaN",
        "nodeType": "function",
        "props": Object {
          "children": <h4>
            <span />
</h4>,
          "className": "type undefined large-type",
          "color": "secondary",
          "onClick": [Function],
          "pill": true,
          "tag": "span",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "span",
          },
          "type": "h4",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "NaN",
        "nodeType": "function",
        "props": Object {
          "children": <h4>
            <span />
</h4>,
          "className": "type undefined large-type",
          "color": "secondary",
          "onClick": [Function],
          "pill": true,
          "tag": "span",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "span",
          },
          "type": "h4",
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Badge
            className="type undefined large-type"
            color="secondary"
            onClick={[Function]}
            pill={true}
            tag="span"
>
            <h4>
                        <span />
            </h4>
</Badge>,
          <Badge
            className="type undefined large-type"
            color="secondary"
            onClick={[Function]}
            pill={true}
            tag="span"
>
            <h4>
                        <span />
            </h4>
</Badge>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "NaN",
          "nodeType": "function",
          "props": Object {
            "children": <h4>
              <span />
</h4>,
            "className": "type undefined large-type",
            "color": "secondary",
            "onClick": [Function],
            "pill": true,
            "tag": "span",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "span",
            },
            "type": "h4",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "NaN",
          "nodeType": "function",
          "props": Object {
            "children": <h4>
              <span />
</h4>,
            "className": "type undefined large-type",
            "color": "secondary",
            "onClick": [Function],
            "pill": true,
            "tag": "span",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "span",
            },
            "type": "h4",
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
